adoc/chapters/cuda_backend.adoc

```diff
@@ -0,0 +1,598 @@
+
+| [code]#template<backend Backend> +
+device +
+make_device(const backend_input_t<Backend, device> &backendObject);# 
+        | Construct a SYCL `device` from a CUDA device. As the SYCL execution environment for the CUDA backend contains a fixed number of devices that are enumerated via `sycl::device::get_devices()`. Calling this function does not create a new device. Rather it merely creates a `sycl::device` object that is a copy of one of the devices from that enumeration.
+
+| [code]#template<backend Backend> +
+context +
+make_context(const backend_input_t<Backend, context> &backendObject,
+                     const async_handler asyncHandler = {});# 
+        | Create a SYCL `context` from a CUDA context.
+
+| [code]#template<backend Backend> +
+queue +
+make_queue(const backend_input_t<Backend, queue> &backendObject,
+                 const context &targetContext,
+                 const async_handler asyncHandler = {});# 
+        | Create a SYCL `queue` from a CUDA stream. The provided `targetContext` must encapsulate the same CUDA context as the provided CUDA stream.
+
+| [code]#template<backend Backend> +
+event +
+make_event(const backend_input_t<Backend, event> &backendObject,
+                 const context &targetContext);# 
+        | Create a SYCL `event` from a CUDA event.
+
+|====
+
+==== Ownership of native backend objects
+
+The CUDA backend retains ownership of all native CUDA objects obtained through
+the interoperability API, therefore associated SYCL objects must be kept alive
+for the duration of the CUDA work using these native CUDA objects.
+
+When creating a SYCL object from a native CUDA object SYCL does not take
+ownership of the object and it is up to the application to dispose of them when
+appropriate.
+
+[[sub:cuda:kernel_function_interoperability]]
+=== Kernel Function Interoperability
+
+This section describes the kernel function interoperability for the CUDA
+backend.
+
+The CUDA backend supports kernel function interoperability for the `accessor`,
+`local_accessor`, `sampled_image_accessor` and `unsampled_image_accessor`
+classes. These are exposed with `get_native` free template function.
+
+The CUDA backend does not support interoperability for the `device_event` class
+as there's no equivalent in CUDA.
+
+Address spaces in CUDA are associated with variable decorations rather than the
+type, so when pointers are passed as parameters to a function the parameter
+types does not need to be decorated with an address space, instead it's simply a
+raw un-decorated pointer. For this reason the `accessor` and  `local_accessor` 
+classes map to a raw undecorated pointer which can be implemented using the 
+generic address space.
+
+Other kernel function types in CUDA are represented by aliases provided in the
+`sycl::cuda` namespace. These are provided for the `sampled_image_accessor`,
+and `unsampled_image_accessor` classes; `sycl::cuda::texture` and
+`sycl::cuda::surface` respectively.
```

Comment:
@AerialMantis what do you think?

Reply:
That's a good point, we should specify what types these are aliases to. There's actually two options for this as CUDA has both the texture and surface reference API (https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#texture-reference-api, https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#surface-reference-api) and the texture and surface object API (https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#texture-object-api, https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#surface-object-api), though the reference API is deprecated so I think the object API makes more sense, in which case the types would be `cudaTextureObject_t` and `cudaSurfaceObject_t`. The object API is actually not templated like the reference API is, so I think we can remove the template arguments on the aliases in the SYCL CUDA backend.

So it would be something like:

```
namespace sycl::cuda {

using texture = cudaTextureObject_t ;
using surface = cudaSurfaceObject_t ;

}  // namespace sycl::cuda
```

Reply:
Thanks for the explanation, @AerialMantis.  I still don't understand why we have aliases for these two types, though.  Why can't `backend_return_t` simply be an alias directly to `cudaTextureObject_t`.  Why does it help to have the intermediate alias named `sycl::cuda::texture`?